lib/decompress/zstd_decompress.c

```diff
@@ -446,7 +447,7 @@ static size_t ZSTD_decodeFrameHeader(ZSTD_DCtx* dctx, const void* src, size_t he
     RETURN_ERROR_IF(dctx->fParams.dictID && (dctx->dictID != dctx->fParams.dictID),
                     dictionary_wrong, "");
 #endif
-    if (dctx->fParams.checksumFlag) XXH64_reset(&dctx->xxhState, 0);
+    if (dctx->fParams.checksumFlag && !dctx->forceIgnoreChecksum) XXH64_reset(&dctx->xxhState, 0);
```

Comment:
Food for though :
I wonder if it would be better to combine `(dctx->fParams.checksumFlag && !dctx->forceIgnoreChecksum)` into a 3rd parameter, named something like `dctx->validateChecksum`.

There is a small optimization here (one variable access vs 2), but I don't this it matters much.
More importantly, I'm wondering if it would be helpful both for code readability and (later) code modifications, as it seems relatively easy to miss one of those conditions.

The critical point here would be to ensure that `dctx->validateChecksum` is always necessarily initialized correctly, which I believe is probably doable.

Reply:
Just for clarity - do you mean:
1) adding the field `validateChecksum` in the `dctx`, that we can initialize based on `fParams.checksumFlag` and `forceIgnoreChecksum`
or 2) changing the `dctx`  param `forceIgnoreChecksum` to `validateChecksum` and figuring out a way to initialize it properly each time, and typically only calling `validateChecksum` in our logic. But, I'm not totally sure how possible this is yet since we do need to know whether compression specified checksumming in some cases.

My only concern about option 1) is that having three flags to deal with checksumming might seem excessive? But I defer that judgement to you.

Reply:
Faced with these 2 choices, I'm more in favor of option 1, for the following reasons : 

- These variables have different nature and lifetime
   + `forceIgnoreChecksum` is a user-selectable parameter. It must be registered "as is", especially as parameters are "sticky" by default, meaning that it should still mean exactly the same for next frame. If it's tampered by some processing involving other local parameters, this property is lost.
   + `fParams.checksumFlag` is a property of the frame currently decoded. It only depends on input, and is not user-selectable. It's useful in itself for other purposes, and only valid for current frame. This flag will have to be determined again on next frame.
   + `validateChecksum` is combination of the previous 2. It's a processing request : it tells that a checksum exist in the frame and must be validated by the decoding process. In contrast, `forceIgnoreChecksum==0` (default) means that _if_ a checksum exist, it should be validated. This request is only valid for the current frame, and will be determined again on next frame.

There is a 3rd possibility, but I'm not sure if it's applicable : 
- If the combination `(dctx->fParams.checksumFlag && !dctx->forceIgnoreChecksum)` is only useful within one function (`ZSTD_decompressContinue()` typically), then `validateChecksum` could be processed as a local stack variable, instead of being an additional state parameter. `validateChecksum` would be recalculated several times per frame, instead of once, but I'm pretty sure it doesn't matter. More importantly, it would ensure that there is no inconsistency between `validateChecksum` and its constituents (`forceIgnoreChecksum` and `fParams.checksumFlag`).